File: src/router.ts
import type { McpManager } from "./manager.js";
import type { JsonRpcRequest, JsonRpcResponse } from "./types.js";
import { ErrorCodes } from "./types.js";

export class Router {
  constructor(private manager: McpManager) {}

  async handleRequest(request: JsonRpcRequest): Promise<JsonRpcResponse> {
    const id = request.id;

    switch (request.method) {
      case "initialize":
        return this.handleInitialize(id);

      case "notifications/initialized":
        return { jsonrpc: "2.0", id, result: {} };

      case "tools/list":
        return this.handleToolsList(id);

      case "tools/call":
        return this.handleToolsCall(id, request.params as Record<string, unknown> | undefined);

      case "ping":
        return { jsonrpc: "2.0", id, result: {} };

      default:
        return {
          jsonrpc: "2.0",
          id,
          error: {
            code: ErrorCodes.METHOD_NOT_FOUND,
            message: `Method not found: ${request.method}`,
          },
        };
    }
  }

  private handleInitialize(id?: string | number): JsonRpcResponse {
    return {
      jsonrpc: "2.0",
      id,
      result: {
        protocolVersion: "2024-11-05",
        capabilities: {
          tools: { listChanged: false },
        },
        serverInfo: {
          name: "mcp-central",
          version: "0.1.0",
        },
      },
    };
  }

  private handleToolsList(id?: string | number): JsonRpcResponse {
    const tools = this.manager.listAllTools();
    return {
      jsonrpc: "2.0",
      id,
      result: { tools },
    };
  }

  private async handleToolsCall(
    id: string | number | undefined,
    params: Record<string, unknown> | undefined
  ): Promise<JsonRpcResponse> {
    if (!params?.name || typeof params.name !== "string") {
      return {
        jsonrpc: "2.0",
        id,
        error: {
          code: ErrorCodes.INVALID_PARAMS,
          message: "Missing 'name' in tools/call params",
        },
      };
    }

    try {
      const result = await this.manager.callTool(
        params.name,
        (params.arguments as Record<string, unknown>) ?? {}
      );
      return { jsonrpc: "2.0", id, result };
    } catch (e) {
      return {
        jsonrpc: "2.0",
        id,
        error: {
          code: ErrorCodes.INTERNAL_ERROR,
          message: e instanceof Error ? e.message : String(e),
        },
      };
    }
  }
}

File: src/aggregator/router.rs
use crate::client::McpManager;
use crate::protocol::{
    error_codes, InitializeResult, JsonRpcId, JsonRpcRequest, JsonRpcResponse,
    ServerCapabilities, ServerInfo, ToolsCapability, ToolsListResult,
};
use std::sync::Arc;

/// Routes MCP requests to the appropriate backend
pub struct Router {
    manager: Arc<McpManager>,
}

impl Router {
    pub fn new(manager: Arc<McpManager>) -> Self {
        Self { manager }
    }

    /// Handle an incoming JSON-RPC request
    pub async fn handle_request(&self, request: JsonRpcRequest) -> JsonRpcResponse {
        let id = request.id.clone();

        match request.method.as_str() {
            "initialize" => self.handle_initialize(id).await,
            "notifications/initialized" => {
                // Just acknowledge, no response needed for notifications
                JsonRpcResponse::success(id, serde_json::json!({}))
            }
            "tools/list" => self.handle_tools_list(id).await,
            "tools/call" => self.handle_tools_call(id, request.params).await,
            "ping" => JsonRpcResponse::success(id, serde_json::json!({})),
            _ => JsonRpcResponse::error(
                id,
                error_codes::METHOD_NOT_FOUND,
                format!("Method not found: {}", request.method),
            ),
        }
    }

    async fn handle_initialize(&self, id: Option<JsonRpcId>) -> JsonRpcResponse {
        let result = InitializeResult {
            protocol_version: "2024-11-05".to_string(),
            capabilities: ServerCapabilities {
                tools: Some(ToolsCapability { list_changed: false }),
                resources: None,
                prompts: None,
            },
            server_info: ServerInfo {
                name: "mcp-bridge".to_string(),
                version: env!("CARGO_PKG_VERSION").to_string(),
            },
        };

        match serde_json::to_value(&result) {
            Ok(value) => JsonRpcResponse::success(id, value),
            Err(e) => JsonRpcResponse::error(id, error_codes::INTERNAL_ERROR, e.to_string()),
        }
    }

    async fn handle_tools_list(&self, id: Option<JsonRpcId>) -> JsonRpcResponse {
        let tools = self.manager.list_all_tools().await;
        let result = ToolsListResult { tools };

        match serde_json::to_value(&result) {
            Ok(value) => JsonRpcResponse::success(id, value),
            Err(e) => JsonRpcResponse::error(id, error_codes::INTERNAL_ERROR, e.to_string()),
        }
    }

    async fn handle_tools_call(
        &self,
        id: Option<JsonRpcId>,
        params: Option<serde_json::Value>,
    ) -> JsonRpcResponse {
        let params = match params {
            Some(p) => p,
            None => {
                return JsonRpcResponse::error(
                    id,
                    error_codes::INVALID_PARAMS,
                    "Missing params for tools/call",
                )
            }
        };

        let name = match params.get("name").and_then(|n| n.as_str()) {
            Some(n) => n.to_string(),
            None => {
                return JsonRpcResponse::error(
                    id,
                    error_codes::INVALID_PARAMS,
                    "Missing 'name' in tools/call params",
                )
            }
        };

        let arguments = params
            .get("arguments")
            .cloned()
            .unwrap_or_else(|| serde_json::json!({}));

        match self.manager.call_tool(&name, arguments).await {
            Ok(result) => match serde_json::to_value(&result) {
                Ok(value) => JsonRpcResponse::success(id, value),
                Err(e) => JsonRpcResponse::error(id, error_codes::INTERNAL_ERROR, e.to_string()),
            },
            Err(e) => JsonRpcResponse::error(id, error_codes::INTERNAL_ERROR, e.to_string()),
        }
    }
}
